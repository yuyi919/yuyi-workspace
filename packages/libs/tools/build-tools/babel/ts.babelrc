{
  "presets": [
    [
      "@babel/preset-typescript",
      {
        "allowNamespaces": false
      }
    ]
  ],
  "plugins": [
    [
      "const-enum",
      {
        "transform": "constObject"
      }
    ],
    [
      "import",
      {
        "libraryName": "lodash",
        "libraryDirectory": "",
        "camel2DashComponentName": false,
        "style": false
      }
    ]
  ]
}
