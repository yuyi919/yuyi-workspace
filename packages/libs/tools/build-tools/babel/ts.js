/**
 *
 * @param {import("@babel/core").ConfigAPI} api
 * @param {*} param1
 */
module.exports = function (api, { jsx }) {
  api.assertVersion(7);
  const plugins = [];
  switch (jsx) {
    case "inferno":
      plugins.push("babel-plugin-inferno");
      break;
    case "vue-next":
      plugins.push("@vue/babel-plugin-jsx");
      break;
  }
  return {
    presets: [
      [
        require.resolve("@babel/preset-typescript"),
        {
          allowNamespaces: true,
        },
      ],
    ],
    plugins: [
      [
        require.resolve("babel-plugin-const-enum"),
        {
          transform: "constObject",
        },
      ],
      ...plugins,
    ],
  };
};
