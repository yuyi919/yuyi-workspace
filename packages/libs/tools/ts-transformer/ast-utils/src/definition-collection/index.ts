export * as Type from "./type"
