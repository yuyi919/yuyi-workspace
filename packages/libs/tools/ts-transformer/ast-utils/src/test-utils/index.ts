export * from "./spec"
export * from "./compile"

