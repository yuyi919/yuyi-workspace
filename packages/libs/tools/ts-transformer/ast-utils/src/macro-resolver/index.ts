import path from "path";
import ts from "typescript";
import { use } from "../visit";
import { VisitorContext } from "../factory";

/**
 * 
 * @param modulePath 
 * @param callerName 
 * @param options 
 */
export function createMacroFunctionResolver(
  modulePath: string,
  callerName: string,
  options: Pick<VisitorContext<any>, "program" | "transformationContext" | "isDeclarationFile">
) {
  const { program, transformationContext: context } = options;
  const moduleJsPath = modulePath.endsWith(".d.ts")
    ? modulePath.replace(/\.d\.ts$/, ".js")
    : modulePath;
  const typeChecker = program.getTypeChecker();
  const Visitor = use(context, program);
  return {
    isImportOrExport(node: ts.Node): node is ts.ImportDeclaration | ts.ExportDeclaration {
      // const c = ts.isImportDeclaration(node) && node.importClause
      // if (c && !ts.isNamespaceImport(c.namedBindings)) {
      //   c.namedBindings.elements.forEach(el => {
      //     const r = el.symbol.getDeclarations()
      //   })
      // }
      return isRefedImportOrExport(node, callerName, moduleJsPath);
    },
    cleanImportOrExport(node: ts.ImportDeclaration | ts.ExportDeclaration) {
      if (ts.isExportDeclaration(node)) {
        return Visitor.visitExportDeclaration(node, options.isDeclarationFile, void 0, (bind: ts.NamedExports) => {
          const nextElements = bind.elements.filter((node) => node.name.text !== callerName);
          if (nextElements.length !== bind.elements.length) {
            Object.assign(bind, { elements: nextElements });
            // bind = ts.factory.updateNamedExports(bind, nextElements);
          }
          return bind;
        });
      }
      return Visitor.visitImportDeclaration(node, options.isDeclarationFile, void 0, (caluse) => {
        var binds: ts.NamedImportBindings | undefined = node.importClause.namedBindings;
        if (ts.isNamedImports(binds)) {
          const nextElements = binds.elements.filter((node) => node.name.text !== callerName);
          // if (nextElements.length === 0) {
          //   if (!node.importClause.name) {
          //   } else {
          //     binds = undefined;
          //   }
          // } else {
          if (nextElements.length !== binds.elements.length) {
            binds = ts.factory.createNamedImports(nextElements);
          }
          // }
        }
        return Object.assign(caluse, { namedBindings: binds });
      });
    },
    isCallExpression(node: ts.Node) {
      return isRefedCallExpression(node, typeChecker, callerName, modulePath);
    },
    isDecorator(node: ts.Node) {
      return ts.isDecorator(node) && isRefedCallExpression(node.expression, typeChecker, callerName, modulePath);
    },
  };
}

function isRefedImportOrExport(
  node: ts.Node,
  callerName: string,
  moduleAbsolutePath: string
): node is ts.ImportDeclaration | ts.ExportDeclaration {
  if (!ts.isImportDeclaration(node) && !ts.isExportDeclaration(node)) {
    return false;
  }
  if (ts.isExportDeclaration(node) && !node.moduleSpecifier) {
    return (
      ts.isNamedExports(node.exportClause) &&
      node.exportClause.elements.some((i) => i.name.text === callerName)
    );
  }
  const module = (node.moduleSpecifier as ts.StringLiteral).text;
  try {
    return (
      moduleAbsolutePath ===
      (module.startsWith(".")
        ? require.resolve(path.resolve(path.dirname(node.getSourceFile().fileName), module))
        : require.resolve(module))
    );
  } catch (e) {
    return false;
  }
}
function isRefedCallExpression(
  node: ts.Node,
  typeChecker: ts.TypeChecker,
  callerName: string,
  moduleAbsolutePath: string
): node is ts.CallExpression {
  if (!ts.isCallExpression(node)) {
    return false;
  }
  const declaration = typeChecker.getResolvedSignature(node)?.declaration;
  if (
    !declaration ||
    ts.isJSDocSignature(declaration) ||
    declaration.name?.getText() !== callerName
  ) {
    return false;
  }
  try {
    // require.resolve is required to resolve symlink.
    // https://github.com/kimamula/ts-transformer-keys/issues/4#issuecomment-643734716
    return require.resolve(declaration.getSourceFile().fileName) === moduleAbsolutePath;
  } catch {
    // declaration.getSourceFile().fileName may not be in Node.js require stack and require.resolve may result in an error.
    // https://github.com/kimamula/ts-transformer-keys/issues/47
    return false;
  }
}
