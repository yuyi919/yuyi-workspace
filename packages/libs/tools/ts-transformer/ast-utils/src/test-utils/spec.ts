import ts from "typescript";
import { readFileSync } from "fs-extra";
import { createTransformContext } from "../factory/context";
import { CustomTransformerContext } from "../factory/types";

export type Spec = (
  node: ts.Node,
  context: CustomTransformerContext
) => ts.Node | void | undefined | false;
export type SpecFactory = (
  context: CustomTransformerContext
) => (node: ts.Node) => ts.Node | void | undefined | false;

export function spec(spec: Spec, path: string) {
  return testCode(spec, readFileSync(path).toString(), path);
}
// endregion

/* ****************************************************************************************************************** */
// region: Transformer
/* ****************************************************************************************************************** */

export function createTransformer(spec: Spec) {}

export function testCode(spec: Spec, code: string, path = "") {
  const sourceFile = ts.createSourceFile(
    path,
    code,
    ts.ScriptTarget.Latest,
    true,
    ts.ScriptKind.TS
  );
  ts.transform(sourceFile, [
    (transformationContext) => {
      const context = createTransformContext(transformationContext, void 0, {}, void 0, {
        fileNames: [path],
        compilerOptions: transformationContext.getCompilerOptions(),
      });
      return (sourceFile) => {
        function visitor(node: ts.Node) {
          const result = spec(node, context);
          if (result) {
            return result;
          }
          return ts.visitEachChild(node, visitor, transformationContext);
        }
        return ts.visitNode(sourceFile, visitor);
      };
    },
  ]);
}
