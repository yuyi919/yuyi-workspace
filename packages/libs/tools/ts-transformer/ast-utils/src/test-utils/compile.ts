import fs from "fs-extra";
import { join, relative, normalize } from "path";
import ts, {
  CompilerOptions,
  createProgram,
  CustomTransformers,
  ModuleKind,
  ScriptTarget,
  WriteFileCallback,
} from "typescript";
import { createTransformContext, CustomTransformerConfig, CustomTransformerHook, VisitorContext } from "../factory";
import { createHarmonyFactory } from "../factory/utils/harmony-factory";
import { Spec, SpecFactory } from "./spec";
export type CustomCompilerOptions<Config> = {
  throwError?: boolean;
  compilerOptions?: Partial<CompilerOptions>;
  tsConfig?: string;
  ts?: string;
  config?: Config;
};

export function transform<Config>(
  filePath: string,
  transformers: (program: ts.Program, options: CustomCompilerOptions<Config>) => CustomTransformers,
  options: CustomCompilerOptions<Config> = {}
) {
  const result: [string, string][] = [];
  compile(
    [filePath],
    transformers,
    (fileName, fileText) => {
      result.push([fileName, fileText]);
    },
    options
  );
  return result;
}
export function transformSpec<Config extends CustomTransformerConfig>(
  nodeVisitor: CustomTransformerHook<Config>,
  filePath: string,
  options?: CustomCompilerOptions<Config>
) {
  return transform(
    filePath,
    (program) => {
      const transformer: ts.TransformerFactory<ts.SourceFile> = (transformationContext) => {
        const context = createTransformContext<Config>(transformationContext, program);
        return (sourceFile) => {
          const visitorContext: VisitorContext<Config> = {
            ...context,
            sourceFile,
            isDeclarationFile: sourceFile.isDeclarationFile,
            originalSourceFile: (<typeof ts>context.tsInstance).getOriginalSourceFile(sourceFile),
            getVisitor() {
              function visitor(node: ts.Node) {
                const result = nodeVisitor.call(visitorContext, node, visitorContext);
                if (result !== void 0) {
                  return result;
                }
                return ts.visitEachChild(node, visitor, transformationContext);
              }
              return visitor;
            },
            factory: createHarmonyFactory(context),
          };
          return context.tsInstance.visitNode(sourceFile, visitorContext.getVisitor());
        };
      };
      const before = options?.config?.before ?? true
      const after = options?.config?.after || false
      const afterDeclarations = options?.config?.afterDeclarations || false
      console.log(`\nbefore %s\nafter: %s\nafterDeclarations: %s`, before, after, afterDeclarations)
      return {
        before: before && [transformer],
        after: after && [transformer],
        afterDeclarations: afterDeclarations && [transformer],
        // afterDeclarations: [transformer],
      };
    },
    options
  ).map(([path, code]) => [normalize(relative(join(__dirname, ".."), path)), code]);
}

export function compile<Config>(
  filePaths: string[],
  transformers: (program: ts.Program, options: CustomCompilerOptions<Config>) => CustomTransformers,
  writeFileCallback?: WriteFileCallback,
  {
    ts = "typescript",
    compilerOptions = {},
    tsConfig: configPath = "",
    ...options
  }: CustomCompilerOptions<Config> = {}
) {
  if (fs.existsSync(configPath)) {
    try {
      const json = fs.readJSONSync(configPath);
      compilerOptions = Object.assign({}, json.compilerOptions || {}, compilerOptions);
    } catch (error) {}
  }
  const program = (require(ts).createProgram as typeof createProgram)(filePaths, {
    strict: true,
    noEmitOnError: false,
    suppressImplicitAnyIndexErrors: true,
    target: ScriptTarget.ES5,
    esModuleInterop: true,
    module: ModuleKind.CommonJS,
    declaration: true,
    noEmitHelpers: false,
    noUnusedLocals: false,
    importHelpers: true,
    emitDecoratorMetadata: true,
    experimentalDecorators: true,
    ...compilerOptions,
  });
  const r = program.emit(
    undefined,
    writeFileCallback,
    undefined,
    false,
    transformers(program, options)
  );
  if ((!options || options.throwError) && r.emitSkipped) {
    const { diagnostics } = r;
    throw new Error(diagnostics.map((diagnostic) => diagnostic.messageText).join("\n"));
  }
  return r;
}
