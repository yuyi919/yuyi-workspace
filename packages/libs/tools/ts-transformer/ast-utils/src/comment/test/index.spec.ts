import ts from "typescript";
import { extractComment, markBarTokenWithTypeNode } from "..";
import * as definition from "../../definition-collection";
import { createMacroFunctionResolver } from "../../macro-resolver";
import { transformSpec } from "../../test-utils";

describe("test", () => {
  it("testResolve", () => {
    const result = transformSpec(
      (node, context) => {
        const { tsInstance } = context;
        const resolver = createMacroFunctionResolver(
          require.resolve("./testFunction"),
          "CREATE",
          context
        );
        // if (resolver.isImportOrExport(node)) {
        //   return resolver.cleanImportOrExport(node);
        // }
        if (tsInstance.isTypeAliasDeclaration(node) && tsInstance.isUnionTypeNode(node.type)) {
          const types = definition.Type.getBaseType(node.type);
          if (
            types instanceof Array &&
            (types.every((o) => o.name === "literal" && o.type === "number") ||
              types.every((o) => o.name === "literal" && o.type === "string"))
          ) {
            types.forEach((member) => {
              const barTokenNode = markBarTokenWithTypeNode(member.node);
              const { comments, tags } = extractComment(barTokenNode, node.getSourceFile());
              expect({ comments, tags }).toMatchSnapshot();
            });
          }
          return node;
        }
        if (resolver.isCallExpression(node)) {
          const name = node.getText();
          console.log(name, "isRefedCallExpression");
          return ts.factory.createObjectLiteralExpression([]);
        }
      },
      require.resolve("./index"),
      {
        compilerOptions: {
          target: ts.ScriptTarget.ESNext,
          module: ts.ModuleKind.ESNext,
        },
      }
    );
    expect(result).toMatchSnapshot();
  });
});
