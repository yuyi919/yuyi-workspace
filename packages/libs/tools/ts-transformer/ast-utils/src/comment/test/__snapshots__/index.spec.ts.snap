// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test testResolve 1`] = `
Object {
  "comments": "测试0",
  "tags": Array [],
}
`;

exports[`test testResolve 2`] = `
Object {
  "comments": "测试1",
  "tags": Array [],
}
`;

exports[`test testResolve 3`] = `
Object {
  "comments": "测试2",
  "tags": Array [],
}
`;

exports[`test testResolve 4`] = `
Array [
  Array [
    "comment\\\\test\\\\testFunction.js",
    "export function CREATE() {
    return {};
}
export const result = 1;
",
  ],
  Array [
    "comment\\\\test\\\\testFunction.d.ts",
    "export declare function CREATE(): CREATE;
export declare type CREATE = {};
export declare const result = 1;
",
  ],
  Array [
    "comment\\\\test\\\\index.js",
    "import { CREATE } from \\"./testFunction\\";
export const a = {};
",
  ],
  Array [
    "comment\\\\test\\\\index.d.ts",
    "import { CREATE } from \\"./testFunction\\";
export declare type Type = 
/**
 * 测试0
 */
0
/**
 * 测试1
 */
 | 1
/**
 * 测试2
 */
 | 2;
export declare const a: CREATE;
",
  ],
]
`;
